File: Data Structures & Algorithms/is-palindrome/submission-1.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        n = len(s)
        if n <= 1: return True
        low, high = 0, n-1
        while low<high:
            while low<high and (not s[low].isalpha() and not s[low].isnumeric()):
                low += 1
            while low<high and (not s[high].isalpha() and not s[high].isnumeric()):
                high -= 1
            if s[low].lower() != s[high].lower(): return False
            low += 1
            high -= 1
        return True

File: Data Structures & Algorithms/eating-bananas/submission-1.cpp
class Solution {
    int required_h(vector<int>& piles, int k) {
        int ret = 0;
        for (int p : piles) {
            ret += (p+k-1)/k;
        }
        return ret;
    }
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int low = 1;
        int high = *max_element(piles.begin(), piles.end());

        while(low<high) {
            int k = (low+high)/2;
            int x = required_h(piles, k);
            if (x<=h) {
                high = k;
            } else {
                low = k+1;
            }
        }
        return low;
    }
};
